src/content/profile-data.js

```diff
@@ -527,3 +527,285 @@ export function getTracingMarkers(thread, markers) {
+    const depthToPop = lastSeenStackIndexByDepth[depth] === stackIndex ? depth : depth - 1;
+    popStacks(depthToPop, previousDepth, sampleTime);
+    pushStacks(depth, stackIndex, sampleTime);
+    previousDepth = depth;
+  }
+
+  // Pop the remaining stacks
+  const endingTime = thread.samples.time[thread.samples.time.length - 1] + interval;
+  popStacks(-1, previousDepth, endingTime);
+
+  // Collapse platform code into single rows if JS only.
+  if (jsOnly) {
+    collapsePlatformStacks(stackTimingByDepth, thread, funcStackInfo);
+  }
+
+  return stackTimingByDepth;
+}
+
+/**
+ * Given timing information like below, collapse out the platform stacks. In the diagram
+ * "J" represents JavaScript stack frame timing, and "P" Platform stack frame timing.
+ * The timing stack index gets changed to -1 for collapsed platform code.
+ *
+ * JJJJJJJJJJJJJJJJ  --->  JJJJJJJJJJJJJJJJ
+ * PPPPPPPPPPPPPPPP        PPPPPPPPPPPPPPPP
+ *     PPPPPPPPPPPP            JJJJJJJJ
+ *     PPPPPPPP                JJJ  JJJ
+ *     JJJJJJJJ
+ *     JJJ  JJJ
+ *
+ * @param {Object} stackTimingByDepth - Table of stack timings.
+ * @param {Object} thread - The current thread.
+ * @param {Object} funcStackInfo - Info about funcStacks.
+ * @returns {Object} The mutated and collapsed timing information.
+ */
+function collapsePlatformStacks(stackTimingByDepth, thread, funcStackInfo) {
+  const {funcStackTable, stackIndexToFuncStackIndex} = funcStackInfo;
+  function findParentTimingIndex(timingRow, start, end) {
+    for (let i = 0; i < timingRow.stack.length; i++) {
+      if (timingRow.start[i] <= start && timingRow.end[i] >= end) {
+        return i;
+      }
+    }
+    return -1;
+  }
+
+  function findChildTimingIndicesInRange(timingRow, start, end) {
+    const inRange = {
+      index: [],
+      stack: [],
+      start: [],
+      end: [],
+    };
+
+    for (let i = 0; i < timingRow.stack.length; i++) {
+      if (timingRow.start[i] >= start) {
+        if (timingRow.end[i] <= end) {
+          inRange.index.push(i);
+          inRange.stack.push(timingRow.stack[i]);
+          inRange.start.push(timingRow.start[i]);
+          inRange.end.push(timingRow.end[i]);
+        } else {
+          break;
+        }
+      }
+    }
+    return inRange;
+  }
+
+  /**
+   * Walk from the parent to the leaf of the stack, collapsing the stack down one level.
+   *
+   * JJJJJJJJJJJJJJJJ
+   * PPPPPPPPPPPPPPPP
+   *     PPPPPPPPPPPP   <- parent
+   *     [ collapse ]   <- empty (the platform timing that was removed)
+   *     JJJJJJJJ       <- child
+   *  ^  JJJ  JJJ
+   *  |
+   *  collapse down
+   *
+   * @param {Integer} depth - The depth of the empty row.
+   * @param {Integer} timingIndex - The timingIndex of the removed timing.
+   * @param {Number} start - The starting time of the removed timing.
+   * @param {Number} end - The ending time of the removed timing.
+   * @return {undefined}
+   */
```

Comment:
I must say these nested functions confuse me, too. I'm used to them when they're just short helpers and when they make use of the outer function's arguments, but I'm inclined to say that long functions like these should be pulled out, especially if they're complex enough that they need a long comment like this one.